src/nrniv/nrnbbcore_write.cpp

```diff
@@ -202,15 +202,20 @@ static bool corenrn_direct;
 static size_t part1();
 static void part2(const char*);
 
-// accessible from ParallelContext.total_bytes()
-size_t nrnbbcore_write() {
-  corenrn_direct = false;
+// prerequisites for a NEURON model to be transferred to CoreNEURON.
+static void model_ready() {
   if (!use_cachevec) {
-    hoc_execerror("nrnbbcore_write requires cvode.cache_efficient(1)", NULL);
+    hoc_execerror("NEURON model for CoreNEURON requires cvode.cache_efficient(1)", NULL);
   }
   if (tree_changed || v_structure_change || diam_changed) {
-    hoc_execerror("nrnbbcore_write requires the model already be initialized (cf finitialize(...))", NULL);
+    hoc_execerror("NEURON model internal structures for CoreNEURON are out of date. (cf finitialize(...))", NULL);
```

Comment:
what is `cf finitialize(...)` means here?

Reply:
A different API definitely needs to be developed and I think that it's an excellent idea for
pc.solve() to be the common simulation advance method.  Presently, pc.nrncore_run(...), is
more a temporary expedient via the passing of a general launch argument string to the coreneuron side. This kind of complexity only a developer can love. And of course it can be wrapped as nicely as one can wish by python code. Nevertheless, I think we should create an issue for this where we can brainstorm a new API.  In some ways it is CVode that wraps the method selections (not only variable time step specific details but a lot of spike management). ParallelContext over the years has gathered into itself not just all multiprocessor activities but threads and a few spike management details. I think that pc.use_coreneuron=1 is insufficient to be a part of management given the number of selection possibilities (cpu/gpu, memory layout, reporting, ...) and I expect we will end up with a new class analogous to CVode that manages all the possible user selections. It is likely in many cases that the selections can be properly computed by the system itself based on the properties of the model. Whether this class is implemented in Python or c++ or both remains to be considered.  I'm not averse to people(e.g. @salvadord) using it as is as their suggestions for improvement will be invaluable. I hope that the api will be able to settle in broad terms over the next several months. I think we are ready for a major push in this area. I don't know if we can reach the user ideal of having distributions where the user can select "use_gpu = 1" without any further attention, but it is the direction I would like to see us go.

Reply:
"NEURON model internal structures for CoreNEURON are out of date."  I thought a hint about what to do about it was in order. But I agree the "cf finitialize" is not a very good hint. I think in most cases putting the cache_efficient(1) call prior to the existing finitialize(...) is the right way to proceed.  However, given that the present scope of CoreNEURON is for fixed step simulations of spike and gap-junction coupled cells, perhaps I should return to the drawing board and give an error message if extracellular or LinearMechanism or RxD or not fixed step.